File: 2kyu/my_smallest_code_interpreter.js
function brainLuck(code, input){
  var data = [0];
  var dt = {};
  var st = [];
  for(var i = 0; i < code.length; i++) {
    if(code[i] == '[') {
      st.push(i);
    } else if(code[i] == ']') {
      var t = st.pop();
      dt[t] = i;
      dt[i] = t;
    }    
  }
  var dp = 0;
  var ip = 0;
  var output = '';
  var len = code.length;
  var cp = 0;

  while(cp < len) {
    switch(code[cp]) {
    case '>':
      dp++;
      if(typeof data[dp] === 'undefined') {
	data[dp] = 0;
      }
      break;
    case '<':
      dp--;
      if(dp < 0)
	dp = 0;
      break;
    case '+':
      data[dp]++;
      if(data[dp] > 255)
	data[dp] = 0;
      break;
    case '-':
      data[dp]--;
      if(data[dp] < 0)
	data[dp] = 255;
      break;
    case '.':
      output += String.fromCharCode(data[dp]);
      break;
    case ',':
      data[dp] = input[ip++].charCodeAt();
      break;
    case '[':
      if(data[dp] === 0)
	cp = dt[cp];
      break;
    case ']':
      if(data[dp] != 0) 
	cp = dt[cp];
      break;
    default:
      break;
    }

    cp++;
  }
  return output;
}

File: 2kyu/my_smallest_code_interpreter.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <assert.h>

#define SIZE 1024

char* brain_luck(char* code, char* input) {
  char* result = (char*)malloc(sizeof(char) * SIZE);
  char data[SIZE] = { 0 };

  int dp = SIZE / 2;
  int cp = 0;
  int size = 0;
  while (code[cp] != '\0') {
    switch(code[cp]) {
    case '>':
      dp++;
      break;
    case '<':
      dp--;
      break;
    case '+':
      data[dp]++;
      break;
    case '-':
      data[dp]--;
      break;
    case '.':
      result[size++] = data[dp];
      break;
    case ',':
      data[dp] = *(input++);
      break;
    case '[':
      if (data[dp] == 0) {
        int ct = 0;
        while (!(code[cp] == ']' && ct == 1)) {
          if (code[cp] == '[') ct++;
          if (code[cp] == ']') ct--;
          cp++;
        }
      }
      break;
    case ']':
      if (data[dp]) {
        int ct = 0;
        while (!(code[cp] == '[' && ct == 1)) {
          if (code[cp] == ']') ct++;
          if (code[cp] == '[') ct--;
          cp--;
        }
      }
      break;
    }
    cp++;
  }
  result[size] = '\0';
  return result;
}

int main() {
  char *code = ",+[-.,+]";
  char *input = "codewars";
  char *res = brain_luck(code, input);
  puts(res);
  free(res);
  return 0;
}
